### PR TITLE
Implement RFC 57 - Reference Unification

### PR DESCRIPTION
**Pull request by pangratz**
[Link to RFC](https://github.com/emberjs/rfcs/pull/57)

### CODE REVIEW

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import Ember from 'ember';
+// FIXME doesn't work an Model is undefined
+// import Model from '../model';
```
Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: Probably due to circular dependencies which are not supported right now

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import Reference from './reference';
+
+var BelongsToReference = function(store, parentInternalModel, belongsToRelationship) {
+  this._super$constructor(store, parentInternalModel);
+  this.belongsToRelationship = belongsToRelationship;
+  this.type = belongsToRelationship.relationshipMeta.type;
+
+  if (belongsToRelationship.link) {
+    this.remoteType = "link";
```
Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: remoteType could change over time, need to make it a function

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+BelongsToReference.prototype.value = function() {
+  var inverseRecord = this.belongsToRelationship.inverseRecord;
+  return inverseRecord && inverseRecord.record;
+};
+
+BelongsToReference.prototype.load = function() {
+  if (this.remoteType === "id") {
+    return this.store.find(this.type, this.id);
```
Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: We should just delegate to RelationshipState here

Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: belongsTo.getRecord should all of this for us

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  if (this.remoteType === "link") {
+    return this.belongsToRelationship.findLink().then((internalModel) => {
+      return this.value();
+    });
+  }
+};
+
+BelongsToReference.prototype.reload = function() {
+  var record = this.value();
```
Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: Should just add a reload to relationshipState

File: packages/ember-data/lib/system/references/has-many.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  this._super$constructor(store, parentInternalModel);
+  this.hasManyRelationship = hasManyRelationship;
+  this.type = hasManyRelationship.relationshipMeta.type;
+
+  if (hasManyRelationship.link) {
+    this.remoteType = "link";
+    this.link = hasManyRelationship.link;
+  } else {
+    this.remoteType = "ids";
+    this._ids = hasManyRelationship.canonicalState.map(function(internalModel) {
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: canonicalState is server state. What we care about is members(or current state but members is better for now). But do not do this eagerly, it's slow and unneccesary

File: packages/ember-data/lib/system/references/has-many.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  // TODO parent
+  // TODO inverse
+};
+
+HasManyReference.prototype = Ember.create(Reference.prototype);
+HasManyReference.prototype.constructor = HasManyReference;
+HasManyReference.prototype._super$constructor = Reference;
+
+HasManyReference.prototype.ids = function() {
+  return this._ids;
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: This can get out of sync. Need to rerun

File: packages/ember-data/lib/system/references/has-many.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+HasManyReference.prototype.meta = function() {
+  return this.hasManyRelationship.manyArray.meta;
+};
+
+HasManyReference.prototype.push = function(objectOrPromise) {
+  return Ember.RSVP.resolve(objectOrPromise).then((data) => {
+    var internalModels = data.map((payload) => {
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: Unsure whether we should support arrays here, probably, but we should at least support json api payload format where it is in `data:`

File: packages/ember-data/lib/system/references/has-many.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      var record = this.store.push(this.type, payload);
+      return record._internalModel;
+    });
+
+    this.hasManyRelationship.addCanonicalRecords(internalModels);
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: Can't just add, what if things were removed. Should call https://github.com/emberjs/data/blob/master/packages/ember-data/lib/system/relationships/state/has-many.js#L122

File: packages/ember-data/lib/system/references/has-many.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    return this.hasManyRelationship.manyArray;
+  });
+};
+
+HasManyReference.prototype._isLoaded = function() {
+  var manyArray = this.hasManyRelationship.manyArray;
+
+  var hasData = get(this.hasManyRelationship, 'hasData');
+  var isEveryLoaded = manyArray.everyProperty('isLoaded', true);
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: Use members, instead of manyArray, we want to make manyArray lazy really soon

File: packages/ember-data/lib/system/references/has-many.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  if (this._isLoaded()) {
+    return this.hasManyRelationship.manyArray;
+  }
+
+  return null;
+};
+
+HasManyReference.prototype.load = function() {
+  if (!this._isLoaded()) {
+    return this.hasManyRelationship.getRecords().get("promise");
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: whats up with get("promise"). Look at my comments for belongsTo

File: packages/ember-data/lib/system/references/has-many.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  var manyArray = this.hasManyRelationship.manyArray;
+  return Ember.RSVP.resolve(manyArray);
+};
+
+HasManyReference.prototype.reload = function() {
+  if (this._isLoaded()) {
+    return this.hasManyRelationship.manyArray.reload();
```
Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: This breaks if it is a `link`. Should use a relationship reload

Comment (igorT) [inline packages/ember-data/lib/system/references/has-many.js]: One of the main reasons we are adding references is to differentiate between a manyArray that has records, [1,2,3] and the relationship, which we might want to reload from `/records/relationship` which returns record [3,4,5]

File: packages/ember-data/lib/system/references/record.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import Ember from 'ember';
+import Reference from './reference';
+
+var RecordReference = function(store, internalModel) {
```
Comment (igorT) [inline packages/ember-data/lib/system/references/record.js]: Not needed for now

File: packages/ember-data/lib/system/store.js [hunk 13]
```diff
@@ -1278,6 +1280,24 @@ Store = Service.extend({
...
   },
 
+  // FIXME move to init
+  setupRecordReferencesCache: Ember.on('init', function() {
+    this.recordReferencesCache = {};
+  }),
+
+  getReference: function(type, id) {
+    var key = `${type}_${id}`;
+    var reference = this.recordReferencesCache[key];
```
Comment (igorT) [inline packages/ember-data/lib/system/store.js]: No reason for a separate cache, can just look it up from internal model

File: packages/ember-data/lib/system/references/belongs-to.js (hunk 2 above)
Comment (pangratz) [inline packages/ember-data/lib/system/references/belongs-to.js]: So `link` and `id` should be a function as well ...

File: packages/ember-data/lib/system/references/belongs-to.js (hunk 4 above)
Comment (pangratz) [inline packages/ember-data/lib/system/references/belongs-to.js]: Should I do this in a separate PR?

Comment (igorT) [inline packages/ember-data/lib/system/references/belongs-to.js]: No, becasue this code is incorrect, so it's easier to make it correct by adding reload onto relationship than hacking here

File: packages/ember-data/lib/system/model/model.js [hunk 14]
```diff
@@ -28,6 +29,15 @@ var retrieveFromCurrentState = Ember.computed('currentState', function(key) {
...
 */
 var Model = Ember.Object.extend(Ember.Evented, {
+
+  init: function() {
+    this._super(...arguments);
+
+    // caches for the relationship references
+    this.belongsToReferences = {};
+    this.hasManyReferences = {};
+  },
```
Comment (pangratz) [inline packages/ember-data/lib/system/model/model.js]: Should the references be stored on the relationship states?

Comment (igorT) [inline packages/ember-data/lib/system/model/model.js]: No

Comment (igorT) [inline packages/ember-data/lib/system/model/model.js]: Not sure actually, keep them on internal model for now

Comment (pangratz) [inline packages/ember-data/lib/system/model/model.js]: Ups, too late. Just pushed. Will revert ...

Comment (pangratz) [inline packages/ember-data/lib/system/model/model.js]: Will move to internal model

File: packages/ember-data/lib/system/model/model.js [hunk 15]
```diff
@@ -742,12 +742,12 @@ var Model = Ember.Object.extend(Ember.Evented, {
     Ember.assert("The `attr` method is not available on DS.Model, a DS.Snapshot was probably expected. Are you passing a DS.Model instead of a DS.Snapshot to your serializer?", false);
   },
 
-  belongsTo: function() {
-    Ember.assert("The `belongsTo` method is not available on DS.Model, a DS.Snapshot was probably expected. Are you passing a DS.Model instead of a DS.Snapshot to your serializer?", false);
+  belongsTo: function(name) {
```
Comment (pangratz) [inline packages/ember-data/lib/system/model/model.js]: TODO add docs

File: packages/ember-data/lib/system/model/model.js [hunk 16]
```diff
@@ -742,12 +742,12 @@ var Model = Ember.Object.extend(Ember.Evented, {
...
+    return this._internalModel.referenceFor('belongsTo', name);
   },
 
-  hasMany: function() {
-    Ember.assert("The `hasMany` method is not available on DS.Model, a DS.Snapshot was probably expected. Are you passing a DS.Model instead of a DS.Snapshot to your serializer?", false);
+  hasMany: function(name) {
```
Comment (pangratz) [inline packages/ember-data/lib/system/model/model.js]: TODO add docs

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+// import Model from '../model';
+import Reference from './reference';
+
+var BelongsToReference = function(store, parentInternalModel, belongsToRelationship) {
+  this._super$constructor(store, parentInternalModel);
+  this.belongsToRelationship = belongsToRelationship;
+  this.type = belongsToRelationship.relationshipMeta.type;
+  this.parent = parentInternalModel.recordReference;
+
+  // TODO inverse
```
Comment (pangratz) [inline packages/ember-data/lib/system/references/belongs-to.js]: @igorT there won't be an inverse, right?

File: packages/ember-data/lib/system/references/belongs-to.js [hunk 18]
```diff
@@ -0,0 +1,87 @@
+import Ember from 'ember';
+// FIXME doesn't work an Model is undefined
+// import Model from '../model';
+import Reference from './reference';
+
+var BelongsToReference = function(store, parentInternalModel, belongsToRelationship) {
+  this._super$constructor(store, parentInternalModel);
+  this.belongsToRelationship = belongsToRelationship;
+  this.type = belongsToRelationship.relationshipMeta.type;
+  this.parent = parentInternalModel.recordReference;
```
Comment (pangratz) [inline packages/ember-data/lib/system/references/belongs-to.js]: Though this doesn't change, should this be a function instead?

File: packages/ember-data/lib/system/store.js [hunk 19]
```diff
@@ -1335,6 +1335,10 @@ Store = Service.extend({
     this.recordArrayManager.recordDidChange(internalModel);
   },
 
+  getReference: function(type, id) {
```
Comment (pangratz) [inline packages/ember-data/lib/system/store.js]: TODO add docs

**Comment by jcope2013** [comment]
status on this?

**Comment by bmac** [comment]
@pangratz do you mind rebasing this pr and adding feature flags so this can be merged?

**Comment by pangratz** [comment]
@bmac done...

**Comment by bmac** [comment]
:tada:

**Comment by sandstrom** [comment]
:balloon: :fireworks:

**Comment by workmanw** [comment]
Wow. I was just catching up on some things and came across this (and the RFC). :+1:. This will definitely help us out in our app. Thank you!
